nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,138 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(%{shape: {_m, k} = shape_q} = q, %{shape: {k, n}} = r, opts) do
+    mode = opts[:mode]
+
+    keep_reflectors = mode == :raw
+    calculate_q = mode in [:reduced, :complete, :raw]
+
+    q =
+      if calculate_q do
+        identity(q, shape: shape_q)
+      end
+
+    {q, r, reflectors} =
+      for i <- 0..(n - 1), reduce: {q, r, []} do
+        {q, r, reflectors} ->
+          %{shape: {reflector_n, reflector_m}} =
+            reflector = householder_reflector(r[[i..(k - 1), i]])
+
+          shape_h = {k, k}
+          h = identity(as_type(reflector, r), shape: shape_h)
+
+          h_bin =
+            for col <- 0..(reflector_m - 1), row <- 0..(reflector_n - 1), reduce: to_binary(h) do
+              acc ->
+                set_value_at_index(
+                  acc,
+                  h.type,
+                  shape_h,
+                  [row + i, col + i],
+                  reflector[[row, col]]
+                )
+            end
```

Comment:
I think this is all going to end up being really expensive. Is there a way you can turn the algorithm into an aggregation/reduction rather than building the binary AOT?

Reply:
This may be possible, although I'll have to sit on this a little bit longer since I'm not _that_ familiar with the algorithm. I'll also keep in mind approaching QR factorization from the Gram-Schmidt and Given Rotations algorithms as well, if Householder ends up leading nowhere.

Reply:
If Household relies on a lot of in place updates, then it's probably not going to be the most efficient at least for the Binary backend.